### PR TITLE
Add support for options --kdistinct and --ordering

### PR DESCRIPTION
**Pull request by mogproject**
- Added options to the following modules:
  - `main.py`
  - `exp_runkernel.py`
  - `exp_runbsd.py`
  - I did not update `exp_memtests.py` because we are not going to use it.

### CODE REVIEW

File: src/algs/kernel.py [hunk 1]
```diff
@@ -4,6 +4,18 @@
 
 import numpy as np
 import algs.utils as utils
+from enum import Enum
+
+
+class OrderingStrategy(Enum):
```
Comment (mogproject) [inline src/algs/kernel.py:10]: Defines ordering strategies in kernelization.

File: src/algs/kernel.py [hunk 2]
```diff
@@ -105,16 +125,22 @@ def remove_and_reorder(A, blocks, k, kernel_v2_enabled):
...
+            permutation = [i for i in A_red.index if i not in representatives] + representatives
+        else:
+            assert False, 'never happens'
         A_red = A_red.reindex(index=permutation, columns=permutation)
 
     return A_red, toremove, num_reduced_blocks
 
 
-def reduction_rules(A, k, kernel_v2_enabled, reorder_vertices=True):
+def reduction_rules(A, k, kernel_v2_enabled, ordering_strategy=OrderingStrategy.PUSH_FRONT):
```
Comment (mogproject) [inline src/algs/kernel.py:143]: Default rule is PUSH_FRONT.

File: src/exp_runbsd.py [hunk 3]
```diff
@@ -378,9 +378,12 @@ def run(pp_dirname, kern_dirname, out_dirname, first_seed, last_seed, k_distinct
         k_distinct = pp_output['post_preprocessing']['kdistinct']
         
         print('ktotal ', k_total, ' kdistinct', k_distinct)
-        if k_distinct>k_distinct_max:
+        if target_kdistinct is None and k_distinct>k_distinct_max:
             print('k_distinct > k_distinct_max....skipping for now')
             continue
+        if target_kdistinct is not None and k_distinct != target_kdistinct:
+            print('skipping because kdistinct != target_kdistinct')
+            continue
```
Comment (mogproject) [inline src/exp_runbsd.py:386]: Note that `k_distinct` here is the minimum number of distinct cliques after preprocessing.

File: src/exp_runbsd.py [hunk 4]
```diff
@@ -463,7 +467,10 @@ def run(pp_dirname, kern_dirname, out_dirname, first_seed, last_seed, k_distinct
                                                                       vertex_indices, groundt_cliqs)
         
         for guess, torun in run_lps.items():
-            if torun:
+            # if the kdistinct value is specified, always enable guesses
+            if target_kdistinct is not None or torun:
```
Comment (mogproject) [inline src/exp_runbsd.py:471]: By default, guesses (trying k-values other than the ground truth k) are disabled if `kdistinct` > 7. But we want to do guesses when `--kdistinct` is specified.

File: src/exp_runbsd.py [hunk 5]
```diff
@@ -505,10 +513,14 @@ def main():
         help="stop BSD after timeout number of seconds", required=False)    
 
     # since v2.0.0
+    parser.add_argument('--kdistinct', type=int,
+        help='limit experiments to the instances with the specific kdistinct value', required=False)
```
Comment (mogproject) [inline src/exp_runbsd.py:517]: e.g. `--kdistinct 13` => run experiments only with the instances with ground-truth kdistinct (after preprocessing) =13.

File: src/exp_runbsd.py [hunk 6]
```diff
@@ -505,10 +513,14 @@ def main():
...
     parser.add_argument('--lp-only', action='store_true',
         help='run algorithm lp only (default: False)')
     parser.add_argument('--kernel-v2', action='store_true',
         help='use kernels created with the v2 option (default: False)')
+    parser.add_argument('--ordering', choices=['arbitrary', 'keep_first', 'keep_last', 'push_front', 'push_back'],
+        help='vertex reordering strategy (default: push_front)', default = 'push_front')
```
Comment (mogproject) [inline src/exp_runbsd.py:523]: e.g. `--ordering arbitrary`, `--ordering keep_first`, etc.

File: src/exp_runbsd.py [hunk 7]
```diff
@@ -529,21 +549,25 @@ def main():
...
+    kernel_dir_suffix = '_v2' if kernel_v2_enabled else ''
+    ordering_suffix = {
+        kernel.OrderingStrategy.ARBITRARY: '_arb',
+        kernel.OrderingStrategy.KEEP_FIRST: '_first',
+        kernel.OrderingStrategy.KEEP_LAST: '_last',
+        kernel.OrderingStrategy.PUSH_FRONT: '',  # default 
+        kernel.OrderingStrategy.PUSH_BACK: '_back',
+    }[ordering_strategy]
+    tf_kern_dirname = f'data/kernels{kernel_dir_suffix}{ordering_suffix}/tf/'
+    lv_kern_dirname = f'data/kernels{kernel_dir_suffix}{ordering_suffix}/lv/'
```
Comment (mogproject) [inline src/exp_runbsd.py:561]: Make sure that for each configuration, the output directory is distinct from the others.

File: src/exp_runbsd.py [hunk 8]
```diff
@@ -485,6 +492,7 @@ def run(pp_dirname, kern_dirname, out_dirname, first_seed, last_seed, k_distinct
 
         # options for v2.0.0
         decomp_output['kernel_v2'] = kernel_v2_enabled
+        decomp_output['ordering'] = ordering_strategy
```
Comment (mogproject) [inline src/exp_runbsd.py:495]: Added new information to the final output.

File: tests/algs/kernel_test.py [hunk 9]
```diff
@@ -27,6 +28,56 @@ def verify_adj_matrix(self, A, A_red, expected_blocks, k, v2):
     def test_reduction_rules(self):
         """Tests reduction_rules()."""
 
+        # tests vertex ordering
```
Comment (mogproject) [inline tests/algs/kernel_test.py:31]: This test covers all ordering strategies.

**Comment by sjain12** [comment]
Looks good.

**Pull request merged**